traits/trait_dict_object.py

```diff
@@ -62,8 +56,299 @@ def __init__(self, added=None, changed=None, removed=None):
+
+        if key_validator is not None:
+            self.key_validator = key_validator
+
+        if value_validator is not None:
+            self.value_validator = value_validator
+
+        if notifiers is None:
+            notifiers = []
+        self.notifiers = list(notifiers)
+
+        if value is None:
+            value = {}
+
+        value = {self.key_validator(key): self.value_validator(value)
+                 for key, value in value.items()}
+
+        super().__init__(value)
+
+    def notify(self, added, changed, removed):
+        """ Call all notifiers.
+
+        This simply calls all notifiers provided by the class, if any.
+        The notifiers are expected to have the signature::
+
+            notifier(trait_dict, added, changed, removed)
+
+        Any return values are ignored.
+
+        """
+
+        for notifier in self.notifiers:
+            notifier(self, added, changed, removed)
+
+    # -- dict interface -------------------------------------------------------
+
+    def __setitem__(self, key, value):
+        """ Set a value for the key, implements self[key] = value.
+
+        Parameters
+        ----------
+        key : A hashable type.
+            The key for the value.
+        value : any
+            The value to set for the corresponding key.
+
+        """
+
+        removed = {}
+        validated_key = self.key_validator(key)
+        validated_value = self.value_validator(value)
+
+        if key in self:
+            changed = {key: self[key]}
+            added = {}
+        else:
+            changed = {}
+            added = {validated_key: validated_value}
+
+        super().__setitem__(validated_key, validated_value)
+        self.notify(added, changed, removed)
+
+    def __delitem__(self, key):
+        """ Delete the item from the dict indicated by the key.
+
+        Parameters
+        ----------
+        key : A hashable type.
+            The key to be deleted.
+
+        Raises
+        ------
+        KeyError
+            If the key is not found.
+
+
+        """
+        removed = {key: self[key]}
+        super().__delitem__(key)
+        self.notify(added={}, changed={}, removed=removed)
+
+    def clear(self):
+        """ Remove all items from the dict. """
+        if self != {}:
+            removed = self.copy()
+            super().clear()
+            self.notify(added={}, changed={}, removed=removed)
+
+    def update(self, adict):
+        """ Update the values in the dict by the new dict.
+
+        Parameters
+        ----------
+        adict : dict
+            The dict from which values will be updated into this dict.
+        """
+
+        validated_dict = {}
+        added = {}
+        changed = {}
+
+        for key, value in adict.items():
+            validated_key = self.key_validator(key)
+            validated_value = self.value_validator(value)
+
+            if key in self:
+                changed[key] = self[key]
```

Comment:
My apologies I overlooked this in the second pass following a change addressing a comment in the first review: this branch became inconsistent with the one below. In the event when the key_validator transform the value, the notification is not consistent with the change on the dict.

Could you please verify this test demonstrates the expected behaviour? If so, please add the test and verify it fails before the fix:
```
    def test_update_with_tranformation(self):

        td = TraitDict(
            {"1": 1, "2": 2},
            key_validator=str,
            notifiers=[self.notification_handler],
        )

        # when
        td.update({1: 2})

        # then
        self.assertEqual(td, {"1": 2, "2": 2})
        self.assertEqual(self.added, {})
        self.assertEqual(self.changed, {"1": 1})
        self.assertEqual(self.removed, {})

```

Then changing these two lines to:
```
            if validated_key in self:
                changed[validated_key] = self[validated_key]
```
should fix the test.

Reply:
Verified that this test would fail before this fix. Thanks!